File: src/frontend/lib/user/models/im/notification_im.dart
import 'watched_item_type_im.dart';

class NotificationIm {
  final DateTime updateTimestamp;
  final WatchedItemTypeIm itemType;
  final String itemId;
  final int itemUpdateCategory;

  NotificationIm({
    required this.updateTimestamp,
    required this.itemType,
    required this.itemId,
    required this.itemUpdateCategory,
  });

  Map<String, dynamic> toJson() => {
        'updateTimestamp': updateTimestamp.millisecondsSinceEpoch,
        'itemType': itemType.index,
        'itemId': itemId,
        'itemUpdateCategory': itemUpdateCategory,
      };
}

File: src/Orchestrator/src/Application/User/Commands/MarkNotificationsAsRead/NotificationIm.cs
using FluentValidation;

using Domain.Aggregates;

using Application.User.Common.Models.IM;

namespace Application.User.Commands.MarkNotificationsAsRead;

public class NotificationIm
{
    public required WatchedItemTypeIm ItemType { get; init; }
    public required Guid ItemId { get; init; }
    public required int ItemUpdateCategory { get; init; }
    public required long UpdateTimestamp { get; init; }
}

internal class NotificationImValidator : AbstractValidator<NotificationIm>
{
    public NotificationImValidator()
    {
        RuleFor(n => n.ItemType).IsInEnum();
        RuleFor(n => n.ItemId).NotEmpty();
        RuleFor(n => n.ItemUpdateCategory).Must(_beAValidUpdateCategory);
        RuleFor(n => n.UpdateTimestamp).GreaterThan(0);
    }

    private bool _beAValidUpdateCategory(int itemUpdateCategory)
    {
        return
            Enum.IsDefined(typeof(SubjectUpdateCategory), itemUpdateCategory) ||
            Enum.IsDefined(typeof(ThingUpdateCategory), itemUpdateCategory) ||
            Enum.IsDefined(typeof(SettlementProposalUpdateCategory), itemUpdateCategory);
    }
}